hazelcast-jet-s3/src/main/java/com/hazelcast/jet/s3/S3Sinks.java

```diff
@@ -0,0 +1,240 @@
+     *                       string representation
+     * @param charset        the name of the charset to be used when encoding
+     *                       the strings
+     * @param <T>            type of the items the sink accepts
+     */
+    @Nonnull
+    public static <T> Sink<? super T> s3(
+            @Nonnull String bucketName,
+            @Nullable String prefix,
+            @Nonnull SupplierEx<? extends AmazonS3> clientSupplier,
+            @Nonnull FunctionEx<? super T, String> toStringFn,
+            @Nonnull String charset
+
+    ) {
+        return SinkBuilder
+                .sinkBuilder("s3-sink", context ->
+                        new S3Context<>(bucketName, prefix, context.globalProcessorIndex(),
+                                clientSupplier, toStringFn, charset))
+                .<T>receiveFn(S3Context::receive)
+                .flushFn(S3Context::flush)
+                .destroyFn(S3Context::close)
+                .build();
+    }
+
+    private static final class S3Context<T> {
+
+        private static final int DEFAULT_MINIMUM_UPLOAD_PART_SIZE = 5 * MB;
+        private final String bucketName;
+        private final String prefix;
+        private final int processorIndex;
+        private final AmazonS3 s3Client;
+        private final FunctionEx<? super T, String> toStringFn;
+        private final String charsetName;
+
+        private int partCounter;
+        private int fileCounter;
+
+        private StringBuilder buffer = new StringBuilder();
+        private List<PartETag> partETags;
+        private String uploadId;
+
+        private S3Context(
+                String bucketName,
+                String prefix,
+                int processorIndex,
+                SupplierEx<? extends AmazonS3> clientSupplier,
+                FunctionEx<? super T, String> toStringFn,
+                String charsetName) {
+            this.bucketName = bucketName;
+            this.prefix = StringUtil.isNullOrEmptyAfterTrim(prefix) ? "" : prefix;
+            this.processorIndex = processorIndex;
+            this.s3Client = clientSupplier.get();
+            this.toStringFn = toStringFn;
+            this.charsetName = charsetName;
+
+            checkIfBucketExists();
+            initUploadRequest();
+        }
+
+        private void initUploadRequest() {
+            partETags = new ArrayList<>();
+            InitiateMultipartUploadRequest initRequest = new InitiateMultipartUploadRequest(bucketName, key());
+            InitiateMultipartUploadResult initResponse = s3Client.initiateMultipartUpload(initRequest);
+            uploadId = initResponse.getUploadId();
+        }
+
+        private void checkIfBucketExists() {
+            if (!s3Client.doesBucketExistV2(bucketName)) {
+                throw new IllegalArgumentException("Bucket [" + bucketName + "] does not exist");
+            }
+        }
+
+        private void receive(T item) {
+            buffer.append(toStringFn.apply(item))
```

Comment:
wouldn't it be better to create a byte buffer and write it to that directly? now we first write to string buffer, and then copy the whole thing again into a byte buffer